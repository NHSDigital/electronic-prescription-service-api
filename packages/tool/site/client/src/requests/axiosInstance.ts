import axios, {InternalAxiosRequestConfig} from "axios"
import {v4 as uuidv4} from "uuid"

const axiosInstance = axios.create({
  validateStatus: () => true
})

axiosInstance.interceptors.request.use(
  (config: InternalAxiosRequestConfig) => {
    const requestId = uuidv4()
    config.headers["x-correlation-id"] = requestId

    return config
  },
  (error) => {
    return Promise.reject(Error(error))
  }
)

export {axiosInstance}
