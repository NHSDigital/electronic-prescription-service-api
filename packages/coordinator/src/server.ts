import Hapi from "@hapi/hapi"
import routes from "./routes"
import HapiPino from "hapi-pino"
import {isLocal} from "./utils/environment"
import {
  reformatUserErrorsToFhir,
  rejectInvalidProdHeaders,
  switchContentTypeForSmokeTest
} from "./utils/server-extensions"

export const createServer = ({collectLogs}: {collectLogs?: boolean}): Hapi.Server => {
  const server = Hapi.server({
    port: 9000,
    host: "0.0.0.0",
    routes: {
      cors: true, // Won't run as Apigee hosted target without this
      payload: {
        parse: false
      }
    }
  })

  if (collectLogs) {
    routes.forEach((route) => {
      route.options = {
        log: {
          collect: true
        }
      }
    })
  }

  server.route(routes)

  server.ext([
    {type: "onRequest", method: rejectInvalidProdHeaders},
    {type: "onPreResponse", method: reformatUserErrorsToFhir},
    {type: "onPreResponse", method: switchContentTypeForSmokeTest}
  ])

  return server
}

const configureLogging = async (server: Hapi.Server) => {
  return await HapiPino.register(server, {
    // For non-local environments, dont pretty print to avoid spamming logs
    ...(isLocal() && {
      transport: {
        target: "pino-pretty",
        options: {
          colorize: true,
          minimumLevel: process.env.LOG_LEVEL || "info",
          levelFirst: true,
          messageFormat: true,
          timestampKey: "time",
          translateTime: true,
          singleLine: false,
          mkdir: true,
          append: true
        }
      }
    }),
    // Redact Authorization headers, see https://getpino.io/#/docs/redaction
    redact: ["req.headers.authorization"],
    wrapSerializers: false
  })
}

export const init = async (): Promise<void> => {
  const server = createServer({})
  await configureLogging(server)
  await server.start()
  server.log("info", `Server running on ${server.info.uri}`)
}
