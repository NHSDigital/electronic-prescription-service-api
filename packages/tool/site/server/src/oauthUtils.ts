import ClientOAuth2 from "client-oauth2"
import {CONFIG} from "./config"
import {getRegisteredCallbackUrl} from "./routes/helpers"
import Hapi from "@hapi/hapi"
import {URLSearchParams} from "url"
import axios from "axios"
import * as jsonwebtoken from "jsonwebtoken"

//TODO: handle system oauth

function createOAuthClient(): ClientOAuth2 {
  return new ClientOAuth2({
    clientId: CONFIG.apigeeAppClientId,
    clientSecret: CONFIG.apigeeAppClientSecret,
    redirectUri: getRegisteredCallbackUrl("callback"),
    accessTokenUri: `${CONFIG.apigeeEgressHost}/oauth2/token`,
    authorizationUri: `${CONFIG.publicApigeeHost}/oauth2/authorize`,
    body: {
      client_id: CONFIG.apigeeAppClientId,
      client_secret: CONFIG.apigeeAppClientSecret
    }
  })
}

// TODO: refresh token
export async function refreshToken(data: ClientOAuth2.Data): Promise<ClientOAuth2.Token> {
  const oauthClientToken = createOAuthClient().createToken(data)
  const refreshedToken = await oauthClientToken.refresh()
  return refreshedToken
}

export interface OAuthTokenResponse {
  access_token: string
  refresh_token: string
  scope: string
  token_type: string
  expires_in: number
}

export interface UserInfoResponse {
  sub: string,
  nationalrbacaccess: {
    nhsid_nrbac_roles: Array<{
      person_roleid: string
    }>
  }
}

interface CIS2TokenResponse extends OAuthTokenResponse {
  id_token: string
}

export async function getCIS2TokenFromAuthCode(request: Hapi.Request): Promise<CIS2TokenResponse> {
  const authorisationCode = request.query.code

  const bodyParams = new URLSearchParams({
    grant_type: "authorization_code",
    client_id: CONFIG.cis2AppClientId,
    client_secret: CONFIG.cis2AppClientSecret,
    redirect_uri: "https://int.api.service.nhs.uk/eps-api-tool/callback",
    code: authorisationCode
  })
  const axiosCIS2TokenResponse = await axios.post<CIS2TokenResponse>(
    `https://${CONFIG.cis2EgressHost}/openam/oauth2/realms/root/realms/NHSIdentity/realms/Healthcare/access_token`,
    bodyParams
  )

  return axiosCIS2TokenResponse.data
}

export async function exchangeCIS2IdTokenForApigeeAccessToken(idToken: string): Promise<OAuthTokenResponse> {
  const apiKey = CONFIG.apigeeAppClientId
  const privateKey = CONFIG.apigeeAppJWTPrivateKey
  const audience = `${CONFIG.publicApigeeHost}/oauth2/token`
  const keyId = CONFIG.apigeeAppJWTKeyId

  const jwt = jsonwebtoken.sign({}, Buffer.from(privateKey, "base64").toString("utf-8"), {
    algorithm: "RS512",
    issuer: apiKey,
    subject: apiKey,
    audience: audience,
    keyid: keyId,
    expiresIn: 300,
    jwtid: crypto.randomUUID()
  })

  //Token Exchange for OAuth2 Access Token
  const bodyParams = new URLSearchParams({
    subject_token: idToken,
    client_assertion: jwt,
    subject_token_type: "urn:ietf:params:oauth:token-type:id_token",
    client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
    grant_type: "urn:ietf:params:oauth:grant-type:token-exchange"
  })

  // TODO: /token failure
  // eslint-disable-next-line max-len
  const axiosOAuthTokenResponse = await axios.post<OAuthTokenResponse>(`${CONFIG.apigeeEgressHost}/oauth2/token`, bodyParams)
  return axiosOAuthTokenResponse.data
}

// eslint-disable-next-line max-len
export async function getApigeeAccessTokenFromAuthCode(request: Hapi.Request, mock: boolean): Promise<OAuthTokenResponse> {
  // TODO: handle code not present
  const authorisationCode = request.query.code

  // TODO: match state on request

  const bodyParams = new URLSearchParams({
    grant_type: "authorization_code",
    client_id: CONFIG.apigeeAppClientId,
    client_secret: CONFIG.apigeeAppClientSecret,
    redirect_uri: getRegisteredCallbackUrl("callback"),
    code: authorisationCode
  })
  const path = mock ? "oauth2-mock/token" : "oauth2/token"
  // TODO: /token failure
  const axiosOAuthTokenResponse = await axios.post<OAuthTokenResponse>(`${CONFIG.apigeeEgressHost}/${path}`, bodyParams)
  return axiosOAuthTokenResponse.data
}

export function getSelectedRoleFromTokenResponse(tokenResponse: OAuthTokenResponse): string | undefined {
  const decodedToken = jsonwebtoken.decode(tokenResponse.access_token) as jsonwebtoken.JwtPayload

  return decodedToken?.["selected_roleid"] as string | undefined
}

export async function getUserInfoRbacRoleFromCIS2Token(
  request: Hapi.Request,
  cis2Token: CIS2TokenResponse
): Promise<string> {
  const logger = request.logger

  const access_token = cis2Token.access_token
  const id_token = cis2Token.id_token

  logger.info("Fetching userinfo to verify sub claim and get RBAC role")
  const axiosUserInfoResponse = await axios.get<UserInfoResponse>(
    `https://${CONFIG.cis2EgressHost}/openam/oauth2/realms/root/realms/NHSIdentity/realms/Healthcare/userinfo`,
    {
      headers: {
        Authorization: `Bearer ${access_token}`
      }
    }
  )
  logger.info(`Userinfo response: ${JSON.stringify(axiosUserInfoResponse.data)}`)

  // Verify sub claim
  const decodedIdToken = jsonwebtoken.decode(id_token) as jsonwebtoken.JwtPayload
  if (decodedIdToken.sub !== axiosUserInfoResponse.data.sub) {
    throw new Error(
      `sub claim in id_token does not match sub in userinfo response:`
        +`${decodedIdToken.sub} !== ${axiosUserInfoResponse.data.sub}`
    )
  }

  const roles = axiosUserInfoResponse.data.nationalrbacaccess.nhsid_nrbac_roles
  logger.info(`Userinfo roles: ${JSON.stringify(roles)}`)

  if (roles.length === 0) {
    throw new Error("No roles found for user")
  }

  logger.info(`Selecting first role from userinfo roles: ${JSON.stringify(roles[0])}`)
  return roles[0].person_roleid
}
