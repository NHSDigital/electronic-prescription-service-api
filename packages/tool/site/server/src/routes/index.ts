import statusRoutes from "./health/get-status"
import loginRoute from "./auth/login"
import oauthCallbackRoute from "./auth/callback"
import refreshRoute from "./auth/refresh"
import logoutRoute from "./auth/logout"
import configRoutes from "./config/config"
import sessionRoutes from "./state/session"
import editRoutes from "./prescribe/edit"
import resetRoutes from "./prescribe/reset"
import signRoutes from "./sign/uploadSignatures"
import sendRoutes from "./sign/downloadSignatures"
import cancelRoutes from "./prescribe/cancel"
import validatorRoutes from "./validate/validator"
import searchRoutes from "./tracker/tracker"
import releaseRoutes from "./dispense/release"
import returnRoutes from "./dispense/return"
import dispenseRoutes from "./dispense/dispense"
import claimRoutes from "./dispense/claim"
import withdrawRoutes from "./dispense/withdraw"
import comparePrescriptions from "./api/comparePrescriptions"
import convertRoutes from "./api/convert"
import sendPrescriptions from "./api/send"
import doseToTextRoutes from "./dose-to-text"
import {isSandbox} from "../services/environment"
import {CONFIG} from "../config"
import {ServerRoute} from "@hapi/hapi"

const authRoutes: Array<ServerRoute> = [
  loginRoute,
  oauthCallbackRoute,
  refreshRoute,
  logoutRoute
]

const apiRoutes: Array<ServerRoute> = [
  ...comparePrescriptions,
  ...convertRoutes,
  ...sendPrescriptions
]

const stateRoutes: Array<ServerRoute> = [
  ...sessionRoutes
]

const prescribingRoutes: Array<ServerRoute> = [
  ...editRoutes,
  ...resetRoutes,
  ...signRoutes,
  ...sendRoutes,
  ...cancelRoutes
]

const validateRoutes: Array<ServerRoute> = [
  ...validatorRoutes
]

const dispensingRoutes: Array<ServerRoute> = [
  ...releaseRoutes,
  ...returnRoutes,
  ...dispenseRoutes,
  ...claimRoutes,
  ...withdrawRoutes
]

const trackerRoutes: Array<ServerRoute> = [
  ...searchRoutes
]

const healthcheckRoutes: Array<ServerRoute> = [
  ...statusRoutes
]

const routes: Array<ServerRoute> = isSandbox(CONFIG.environment)
  ? [
    ...healthcheckRoutes,
    ...doseToTextRoutes,
    oauthCallbackRoute
  ]
  : [
    ...configRoutes,
    ...authRoutes,
    ...apiRoutes,
    ...stateRoutes,
    ...healthcheckRoutes,
    ...prescribingRoutes,
    ...validateRoutes,
    ...dispensingRoutes,
    ...trackerRoutes,
    ...doseToTextRoutes
  ]

export default routes
