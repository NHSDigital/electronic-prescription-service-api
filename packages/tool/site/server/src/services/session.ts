import Hapi from "@hapi/hapi"
import {getUtcEpochSeconds} from "../routes/util"
import {CONFIG} from "../config"
import {isLocal} from "./environment"
import {OAuthTokenResponse} from "../oauthUtils"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getSessionValue(key: string, request: Hapi.Request): any {
  const sessionValue = request.yar.get(key)
  if (isLocal(CONFIG.environment)) {
    if (sessionValue === null) {
      request.logger.error(`Failed to retrieve session value for key: ${key}`)
    } else {
      request.logger.info(`Retrieved ${key} from session with value: ${JSON.stringify(sessionValue)}`)
    }
  }
  if (sessionValue && Object.keys(sessionValue).length === 1 && Object.keys(sessionValue)[0] === "arrayValues") {
    return sessionValue.arrayValues
  }
  return sessionValue
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getSessionValueOrDefault(key: string, request: Hapi.Request, fallback: unknown): any {
  return getSessionValue(key, request) ?? fallback
}

export function setSessionValue(key: string, value: unknown, request: Hapi.Request): void {
  if (Array.isArray(value)) {
    value = {arrayValues: value}
  }
  if (isLocal(CONFIG.environment)) {
    request.logger.info(`Saving ${key} to session with value: ${JSON.stringify(value)}`)
  }
  request.yar.set(key, value)
}

export function appendToSessionValue(key: string, value: unknown, request: Hapi.Request): void {
  const existingValue = getSessionValueOrDefault(key, request, [])
  if (!Array.isArray(existingValue)) {
    throw Error(`Cannot append to session value with key: '${key}', session value is not an array`)
  }
  const mergedValue = existingValue.concat(value)
  setSessionValue(key, mergedValue, request)
}

export function appendToSessionValueWithoutDuplication(key: string, value: unknown, request: Hapi.Request): void {
  const existingValue = getSessionValueOrDefault(key, request, [])
  if (!Array.isArray(existingValue)) {
    throw Error(`Cannot append to session value with key: '${key}', session value is not an array`)
  }
  if (!existingValue.includes(value)) {
    const mergedValue = existingValue.concat(value)
    setSessionValue(key, mergedValue, request)
  }
}

export function removeFromSessionValue(key: string, value: unknown, request: Hapi.Request): void {
  const existingValue = getSessionValueOrDefault(key, request, [])
  if (!Array.isArray(existingValue)) {
    throw Error(`Cannot remove an entry from session value with key: '${key}', session value is not an array`)
  }
  const valueWithEntryRemoved = existingValue.filter(v => v !== value)
  setSessionValue(key, valueWithEntryRemoved, request)
}

export function clearSessionValue(key: string, request: Hapi.Request): void {
  request.yar.clear(key)
}

function createAuthSession(tokenResponse: OAuthTokenResponse, request: Hapi.Request, h: Hapi.ResponseToolkit) {
  request.cookieAuth.set({})
  const accessTokenFetchTime = getUtcEpochSeconds()
  const refreshTokenTimeout = CONFIG.refreshTokenTimeout
  const timeTillRefresh = 599
  const nextRefreshTime = accessTokenFetchTime + timeTillRefresh - 10
  request.cookieAuth.ttl(refreshTokenTimeout)
  // TODO: type session values
  setSessionValue("access_token_data", tokenResponse, request)
  setSessionValue("next_refresh_time", nextRefreshTime, request)
  h.state("Next-Refresh-Time", nextRefreshTime.toString(), {isHttpOnly: false})
  h.state("Access-Token-Fetched", accessTokenFetchTime.toString(), {isHttpOnly: false})
  h.state("Access-Token-Set", "true", {isHttpOnly: false, ttl: refreshTokenTimeout})
  h.state("Token-Expires-In", (refreshTokenTimeout / 1000).toString(), {isHttpOnly: false})
}

export function createSandboxAuthSession(request: Hapi.Request, h: Hapi.ResponseToolkit): void {
  const sandboxTokenResponse: OAuthTokenResponse = {
    access_token: "sandboxAccessToken",
    refresh_token: "sandboxRefreshToken",
    scope: "",
    token_type: "Bearer",
    expires_in: 1000 * 60 * 60
  }
  createAuthSession(sandboxTokenResponse, request, h)
}

export function createCombinedAuthSession(
  tokenResponse: OAuthTokenResponse, request: Hapi.Request, h: Hapi.ResponseToolkit
): void {
  createAuthSession(tokenResponse, request, h)
  h.state("Auth-Method", "Combined")
  h.state("Auth-Level", "User")
}

export function createSeparateAuthSession(
  tokenResponse: OAuthTokenResponse, request: Hapi.Request, h: Hapi.ResponseToolkit
): void {
  createAuthSession(tokenResponse, request, h)
  h.state("Auth-Method", "Separate")
  h.state("Auth-Level", "User")
}

export function getApigeeAccessTokenFromSession(request: Hapi.Request): string {
  if (isLocal(CONFIG.environment)) {
    return "sandbox-token"
  }
  const accessTokenData = getSessionValue("access_token_data", request)
  return accessTokenData.access_token
}

export function clearSession(request: Hapi.Request, h: Hapi.ResponseToolkit): void {
  request.yar.reset()
  request.cookieAuth.clear()
  h.unstate("Next-Refresh-Time")
  h.unstate("Access-Token-Fetched")
  h.unstate("Access-Token-Set")
  h.unstate("Token-Expires-In")
  h.unstate("Auth-Method")
  h.unstate("Auth-Level")
}
