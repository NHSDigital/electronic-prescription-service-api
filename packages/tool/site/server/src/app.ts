import Hapi, {RouteDefMethods, RouteOptions} from "@hapi/hapi"
import routes from "./routes"
import HapiPino from "hapi-pino"
import Vision from "@hapi/vision"
import * as inert from "@hapi/inert"
import Yar from "@hapi/yar"
import Cookie from "@hapi/cookie"
import {isLocal, isSandbox} from "./services/environment"
import axios from "axios"
import {CONFIG} from "./config"
import {getSessionValue} from "./services/session"
import * as XLSX from "xlsx"

const init = async () => {
  axios.defaults.validateStatus = () => true

  const server = createServer()

  await registerAuthentication(server)
  await registerSession(server)
  await registerLogging(server)
  await registerStaticRouteHandlers(server)
  await registerViewRouteHandlers(server)

  addStaticRoutes(server)
  addDownloadRoutes(server)
  addApiRoutes(server)
  addViewRoutes(server)

  await server.start()
  server.log("info", `Server running on ${server.info.uri}`)
}

process.on("unhandledRejection", err => {
  console.log(err)
  process.exit(1)
})

function createServer() {
  return Hapi.server({
    port: 9000,
    host: "0.0.0.0",
    routes: {
      cors: true
    }
  })
}

async function registerAuthentication(server: Hapi.Server) {
  await server.register(Cookie)
  server.auth.strategy("session", "cookie", {
    cookie: {
      name: "auth",
      password: CONFIG.sessionKey,
      isSecure: true
    },
    redirectTo: () => {
      const needsLogin = !(isSandbox(CONFIG.environment) || isLocal(CONFIG.environment))
      return needsLogin ? `${CONFIG.baseUrl}login` : `${CONFIG.baseUrl}callback`
    }
  })
  server.auth.default("session")
}

async function registerSession(server: Hapi.Server) {
  await server.register({
    plugin: Yar,
    options: {
      storeBlank: true,
      // Use "0" maxCookieSize to force all session data to be written to cache
      maxCookieSize: 0,
      cache: {
        expiresIn: 24 * 60 * 60 * 1000
      },
      cookieOptions: {
        password: CONFIG.sessionKey,
        isSecure: true,
        isSameSite: "None"
      }
    }
  })
}

async function registerLogging(server: Hapi.Server) {
  await HapiPino.register(server, {
    // For non-local environments, dont pretty print to avoid spamming logs
    ...(isLocal(CONFIG.environment) && {
      transport: {
        target: "pino-pretty",
        options: {
          colorize: true,
          minimumLevel: "info",
          levelFirst: true,
          messageFormat: true,
          timestampKey: "time",
          translateTime: true,
          singleLine: false,
          mkdir: true,
          append: true
        }
      }
    }),
    // Redact Authorization headers, see https://getpino.io/#/docs/redaction
    redact: ["req.headers.authorization"],
    wrapSerializers: false
  })
}

async function registerStaticRouteHandlers(server: Hapi.Server) {
  await server.register(inert)
}

async function registerViewRouteHandlers(server: Hapi.Server) {
  await server.register(Vision)
  server.views({
    engines: {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      html: require("handlebars")
    },
    relativeTo: __dirname,
    path: "templates"
  })
}

function addStaticRoutes(server: Hapi.Server) {
  server.route({
    method: "GET",
    path: `/static/{param*}`,
    options: {
      auth: false
    },
    handler: {
      directory: {
        path: "static"
      }
    }
  })
}

function addDownloadRoutes(server: Hapi.Server) {
  server.route({
    method: "GET",
    path: "/download/exception-report",
    handler: downloadExceptionReport
  })

  function downloadExceptionReport(request: Hapi.Request, h: Hapi.ResponseToolkit) {
    const exceptions = getSessionValue("exception_report", request)
    const fileName = "exception-report"
    const wb = XLSX.utils.book_new()
    const ws = XLSX.utils.json_to_sheet(exceptions)
    XLSX.utils.book_append_sheet(wb, ws, "Test Exception Report")

    return h
      .response(XLSX.write(wb, {type: "binary"}))
      .type("application/vnd.ms-excel")
      .header("content-disposition", `attachment; filename=${fileName}.xlsx;`)
      .encoding("binary")
      .code(200)
  }
}

function addApiRoutes(server: Hapi.Server) {
  server.route(routes)
}

function addViewRoutes(server: Hapi.Server) {
  server.route(addHomeView())

  if (isSandbox(CONFIG.environment)) {
    server.route(addView("dose-to-text"))
  } else {
    server.route(addView("login", true))
    server.route(addView("my-prescriptions"))
    server.route(addView("validate"))
    server.route(addView("compare-prescriptions"))
    server.route(addView("search"))
    server.route(addView("view"))
    server.route(addView("tracker"))
    server.route(addView("prescribe/load"))
    server.route(addView("prescribe/edit"))
    server.route(addView("prescribe/send", true))
    server.route(addView("prescribe/cancel"))
    server.route(addView("dispense/release"))
    server.route(addView("dispense/verify"))
    server.route(addView("dispense/return"))
    server.route(addView("dispense/dispense"))
    server.route(addView("dispense/withdraw"))
    server.route(addView("dispense/claim"))
  }

  if (isLocal(CONFIG.environment)) {
    server.route(addView("dose-to-text"))
  }

  server.route(addView("config"))

  function addHomeView(): Hapi.ServerRoute {
    return addView("/")
  }

  function addView(path: string, skipAuth?: boolean): Hapi.ServerRoute {
    const viewRoute = {
      method: "GET" as RouteDefMethods,
      path: path.startsWith("/") ? path : `/${path}`,
      handler: {
        view: {
          template: "index",
          context: {
            baseUrl: CONFIG.baseUrl,
            environment: CONFIG.environment
          }
        }
      }
    }

    if (skipAuth) {
      return {
        ...viewRoute,
        options: {
          auth: false
        } satisfies RouteOptions
      }
    }

    return viewRoute
  }
}

init()
