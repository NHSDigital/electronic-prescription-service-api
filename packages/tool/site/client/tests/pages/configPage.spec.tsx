import {waitFor} from "@testing-library/react"
import {screen} from "@testing-library/dom"
import pretty from "pretty"
import * as React from "react"
import MockAdapter from "axios-mock-adapter"
import {AppContextValue} from "../../src"
import {renderWithContext} from "../renderWithContext"
import userEvent from "@testing-library/user-event"
import {axiosInstance} from "../../src/requests/axiosInstance"
import {int, internalDev} from "../../src/services/environment"
import ConfigPage from "../../src/pages/configPage"
import {MemoryRouter} from "react-router-dom"

const baseUrl = "baseUrl/"

const configUrl = `${baseUrl}config`

const mock = new MockAdapter(axiosInstance)

beforeEach(() => mock.reset())
afterEach(() => mock.reset())

test("Displays config form correctly in internal dev", async () => {
  const context: AppContextValue = {baseUrl, environment: internalDev}
  const container = await renderPage(context)

  expect(screen.getByText("Config")).toBeTruthy()
  expect(screen.getByText("EPS PR Number")).toBeTruthy()
  expect(screen.getByText("Use Proxygen deployed APIs")).toBeTruthy()
  expect(screen.getByText("Use Signing Mock")).toBeTruthy()
  expect(screen.getByText("Signing PR Number")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays config form correctly in int", async () => {
  const context: AppContextValue = {baseUrl, environment: int}
  const container = await renderPage(context)

  expect(screen.getByText("Config")).toBeTruthy()
  expect(screen.queryByText("EPS PR Number")).toBeNull()
  expect(screen.getByText("Use Proxygen deployed APIs")).toBeTruthy()
  expect(screen.queryByText("Use Signing Mock")).toBeNull()
  expect(screen.queryByText("Signing PR Number")).toBeNull()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays config update result", async () => {
  const context: AppContextValue = {baseUrl, environment: internalDev}
  mock.onAny(configUrl).reply(200, {
    success: true
  })

  const container = await renderPage(context)
  userEvent.click(screen.getByText("Save"))
  await waitFor(() => screen.getByText(/Config Saved/))
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

async function renderPage(context) {
  const {container} = renderWithContext(<MemoryRouter><ConfigPage/></MemoryRouter>, context)
  await waitFor(() => screen.getByText("Config"))
  return container
}
