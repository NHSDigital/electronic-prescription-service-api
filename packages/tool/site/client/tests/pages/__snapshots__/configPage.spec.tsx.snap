// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays config form correctly in int 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Config</label></h1>
<form><label class="nhsuk-label nhsuk-label--s">EPS</label>
  <div class="nhsuk-form-group">
    <fieldset class="nhsuk-fieldset"><label class="nhsuk-label nhsuk-label--s">Use Proxygen</label>
      <div class="nhsuk-form-group">
        <div class="nhsuk-checkboxes" id="useProxygen">
          <div class="nhsuk-checkboxes__item"><input name="useProxygen" id="useProxygen" type="checkbox" data-checkbox-exclusive-group="useProxygen" value="false"><label class="nhsuk-label nhsuk-checkboxes__label" id="useProxygen--label" for="useProxygen">Use Proxygen deployed APIs</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class="sc-dQkurY gNfCfq"><button class="nhsuk-button" aria-disabled="false" type="submit">Save</button></div>
</form>"
`;

exports[`Displays config form correctly in internal dev 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Config</label></h1>
<form><label class="nhsuk-label nhsuk-label--s">EPS</label>
  <div class="nhsuk-form-group">
    <fieldset class="nhsuk-fieldset">
      <div class="nhsuk-form-group"><label class="nhsuk-label" id="epsPrNumber--label" for="epsPrNumber">EPS PR Number</label><input class="nhsuk-input nhsuk-input--width-30" type="text" name="epsPrNumber" id="epsPrNumber" value=""></div><label class="nhsuk-label nhsuk-label--s">Use Proxygen</label>
      <div class="nhsuk-form-group">
        <div class="nhsuk-checkboxes" id="useProxygen">
          <div class="nhsuk-checkboxes__item"><input name="useProxygen" id="useProxygen" type="checkbox" data-checkbox-exclusive-group="useProxygen" value="false"><label class="nhsuk-label nhsuk-checkboxes__label" id="useProxygen--label" for="useProxygen">Use Proxygen deployed APIs</label></div>
        </div>
      </div><label class="nhsuk-label nhsuk-label--s">Signing</label>
      <div class="nhsuk-form-group">
        <div class="nhsuk-checkboxes" id="useSigningMockCheckboxes">
          <div class="nhsuk-checkboxes__item"><input name="useSigningMock" id="useSigningMock" type="checkbox" data-checkbox-exclusive-group="useSigningMockCheckboxes" value="false"><label class="nhsuk-label nhsuk-checkboxes__label" id="useSigningMock--label" for="useSigningMock">Use Signing Mock</label></div>
        </div>
      </div>
      <div class="nhsuk-form-group"><label class="nhsuk-label" id="signingPrNumber--label" for="signingPrNumber">Signing PR Number</label><input class="nhsuk-input nhsuk-input--width-30" type="text" name="signingPrNumber" id="signingPrNumber" value=""></div>
    </fieldset>
  </div>
  <div class="sc-dQkurY gNfCfq"><button class="nhsuk-button" aria-disabled="false" type="submit">Save</button></div>
</form>"
`;

exports[`Displays config update result 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Config Saved <svg class="nhsuk-icon nhsuk-icon__tick" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" aria-hidden="true" height="34" width="34">
      <path stroke-width="4" stroke-linecap="round" stroke="#007f3b" fill="none" d="M18.4 7.8l-8.5 8.4L5.6 12"></path>
    </svg></label></h1>
<div class="sc-dQkurY gNfCfq"><button class="nhsuk-button nhsuk-button--secondary" aria-disabled="false" type="button">Back</button></div>"
`;
